### PR TITLE
Ghost role time restrictions

### PR DESCRIPTION
**Pull request by VigersRay**
## About the PR

Adds the ability to specify time limits for ghost roles. This is a priority for ERT so that people who do not understand anything in medicine or engineering do not enter them. In the future, you can add restrictions on the roles of antagonists, such as syndicate reinforcements.
It also slightly improves the appearance of the ghost roles window.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

image

**Changelog**

### CODE REVIEW

File: Content.Client/Players/PlayTimeTracking/JobRequirementsManager.cs [hunk 1]
```diff
@@ -117,4 +117,33 @@ public bool IsAllowed(JobPrototype job, [NotNullWhen(false)] out string? reason)
...
+                reasonBuilder.Append('\n');
+            first = false;
+
+            reasonBuilder.AppendLine(reason);
+        }
+
+        reason = reasonBuilder.Length == 0 ? null : reasonBuilder.ToString();
+        return reason == null;
+    }
 }
```
Comment (metalgearsloth) [inline Content.Client/Players/PlayTimeTracking/JobRequirementsManager.cs:149]: Don't copy-paste code between methods, re-use it.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesEui.cs [hunk 2]
```diff
@@ -64,14 +66,26 @@ public override void HandleState(EuiStateBase state)
             if (state is not GhostRolesEuiState ghostState) return;
             _window.ClearEntries();
 
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var sysManager = entityManager.EntitySysManager;
+            var spriteSystem = sysManager.GetEntitySystem<SpriteSystem>();
+            var requirementsManager = IoCManager.Resolve<JobRequirementsManager>();
+
```
Comment (metalgearsloth) [inline Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesEui.cs:73]: These dependencies should be stored on the class itself.

File: Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesEntry.xaml.cs [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
     {
+        private SpriteSystem _spriteSystem;
         public event Action<GhostRoleInfo>? OnRoleSelected;
         public event Action<GhostRoleInfo>? OnRoleFollow;
 
-        public GhostRolesEntry(string name, string description, IEnumerable<GhostRoleInfo> roles)
+        public GhostRolesEntry(string name, string description, bool hasAccess, string? reason, IEnumerable<GhostRoleInfo> roles, SpriteSystem spriteSystem)
         {
             RobustXamlLoader.Load(this);
+            _spriteSystem = spriteSystem;
```
Comment (metalgearsloth) [inline Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesEntry.xaml.cs:21]: Why is spritesystem a field when it's only used in the ctor?

**Comment by metalgearsloth** [comment]
@VigersRay are you coming back to this.

**Comment by VigersRay** [comment]
> @VigersRay are you coming back to this.

I'll do it today or tomorrow.

**Comment by VigersRay** [comment]
> @VigersRay are you coming back to this.

I'll do it today or tomorrow.

**Comment by metalgearsloth** [comment]
re-open when you get a chance

File: Content.Client/Players/PlayTimeTracking/JobRequirementsManager.cs (hunk 1 above)
Comment (Geekyhobo) [reply Content.Client/Players/PlayTimeTracking/JobRequirementsManager.cs:149]: So use IsAllowed?

**Pull request closed without merging**